core/src/connection/manager/task.rs

```diff
@@ -0,0 +1,333 @@
+            Event::Error { id, .. } => id,
+            Event::Notify { id, .. } => id,
+        }
+    }
+}
+
+/// A `Task` is a [`Future`] that handles a single connection.
+pub struct Task<F, M, H, I, O, E, C>
+where
+    M: StreamMuxer,
+    H: IntoConnectionHandler<C>,
+    H::Handler: ConnectionHandler<Substream = Substream<M>>
+{
+    /// The ID of this task.
+    id: TaskId,
+
+    /// Sender to emit events to the manager of this task.
+    events: mpsc::Sender<Event<O, H, E, <H::Handler as ConnectionHandler>::Error, C>>,
+
+    /// Receiver for commands sent by the manager of this task.
+    commands: stream::Fuse<mpsc::Receiver<Command<I>>>,
+
+    /// Inner state of this `Task`.
+    state: State<F, M, H, I, O, E, C>,
+}
+
+impl<F, M, H, I, O, E, C> Task<F, M, H, I, O, E, C>
+where
+    M: StreamMuxer,
+    H: IntoConnectionHandler<C>,
+    H::Handler: ConnectionHandler<Substream = Substream<M>>
+{
+    /// Create a new task to connect and handle some node.
+    pub fn pending(
+        id: TaskId,
+        events: mpsc::Sender<Event<O, H, E, <H::Handler as ConnectionHandler>::Error, C>>,
+        commands: mpsc::Receiver<Command<I>>,
+        future: F,
+        handler: H
+    ) -> Self {
+        Task {
+            id,
+            events,
+            commands: commands.fuse(),
+            state: State::Pending {
+                future: Box::pin(future),
+                handler,
+                events: Vec::new()
+            },
+        }
+    }
+
+    /// Create a task for an existing node we are already connected to.
+    pub fn established(
+        id: TaskId,
+        events: mpsc::Sender<Event<O, H, E, <H::Handler as ConnectionHandler>::Error, C>>,
+        commands: mpsc::Receiver<Command<I>>,
+        connection: Connection<M, H::Handler>
+    ) -> Self {
+        Task {
+            id,
+            events,
+            commands: commands.fuse(),
+            state: State::EstablishedPending(connection),
+        }
+    }
+}
+
+/// The state associated with the `Task` of a connection.
+enum State<F, M, H, I, O, E, C>
+where
+    M: StreamMuxer,
+    H: IntoConnectionHandler<C>,
+    H::Handler: ConnectionHandler<Substream = Substream<M>>
+{
+    /// The task is waiting for the connection to be established.
+    Pending {
+        /// The future that will attempt to reach the node.
+        // TODO: don't pin this Future; this requires deeper changes though
+        future: Pin<Box<F>>,
+        /// The intended handler for the established connection.
+        handler: H,
+        /// While we are dialing the future, we need to buffer the events received via
+        /// `Command::NotifyHandler` so that they get delivered to the `handler`
+        /// once the connection is established. We can't leave these in `Task::receiver`
+        /// because we have to detect if the connection attempt has been aborted (by
+        /// dropping the corresponding `sender` owned by the manager).
+        events: Vec<I>
```

Comment:
Should we limit the amount of events one can buffer here? Or can this not cause any issues, given that the pending connection eventually times out?

Reply:
This is also pre-existing code, so I'm not sure. Since dropping events is out of the question, as it may lead to arbitrary misbehaviour on behalf of the connection handler, the only options seem to be either to establish another mechanism for backpressure towards the manager, which I'm not sure how to do here, or to at least have a hard limit that avoids theoretical unbounded growth and when hit, the task aborts the pending connection and logs an error. This last option sounds the most appealing to me. What do you think?

Reply:
Since this is pre-existing code and the solution isn't obvious, I'd suggest we open an issue and tackle it afterwards.

Reply:
See https://github.com/libp2p/rust-libp2p/issues/1465.